### PR TITLE
DAOS-13774 control: Don't append ofi_rxm to ofi+tcp provider

### PR DESCRIPTION
**Pull request by kjacque**
This change requires ofi_rxm to be specified manually if it is desired for ofi+tcp. There is now a non-RXM tcp provider available in libfabric.

Features: control

Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by knard-intel — DISMISSED** [review]

**Review by daltonbohning — CHANGES_REQUESTED** [review]
-1 until @phender confirms my comment

File: src/tests/ftest/launch.py
Comment (kjacque) [reply src/tests/ftest/launch.py:64]: Hm, for the record the Jenkinsfile is different in the master branch, and the defaultValue string for TestProvider is empty.

Comment (daltonbohning) [reply src/tests/ftest/launch.py:64]: Yeah, the TCP provider testing is a different branch / different Jenkinsfile. That makes updating difficult. Because if we update one branch without updating the other, something will break.

Comment (daltonbohning) [reply src/tests/ftest/launch.py:64]: @phender maybe we could do this?
1. default `ofi+tcp` -> `ofi+tcp;ofi_rxm` in launch.py
2. update the provider testing branch to use `ofi+tcp;ofi_rxm`
3. remove the default `ofi+tcp` -> `ofi+tcp;ofi_rxm` in launch.py

Comment (phender) [reply src/tests/ftest/launch.py:64]: > @phender I think this will break TCP provider testing, right?
> 
> https://github.com/daos-stack/daos/blob/f33fae751e56e6926f69b453ec2de6d5014f0850/Jenkinsfile#L82

It shouldn't.  The `PROVIDER_KEYS.value` is used to ultimately set `CRT_PHY_ADDR_STR`, which is then used here https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/server_utils_params.py#L111 to set the default value for the server config file `provider` entry.

We may need to adjust https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/network_utils.py#L16.

We should test this PR with different `Test-provider: <proivider>` commit pragmas (or build params) and include at least `test_dmg_network_scan_basic` along with the `pr` tag.

Comment (daltonbohning) [reply src/tests/ftest/launch.py:64]: But if `--provider ofi+tcp` is passed, it wouldn't be in `PROVIDER_KEYS`, since `PROVIDER_KEYS` now has `ofi+tcp;ofi_rxm`, but NOT `ofi+tcp`

Comment (phender) [reply src/tests/ftest/launch.py:64]: We'll also need to update the descriptive messages in the Jenkinsfile, e.g. https://github.com/daos-stack/daos/blob/provider-testing-tcp/Jenkinsfile#L83, in the affected branches: `master`, `provider-testing-tcp`, release/2.4`, and `provider-testing-tcp-2.4-testing` (assuming this will be cherry-picked).

Comment (phender) [reply src/tests/ftest/launch.py:64]: > @phender maybe we could do this?
> 
> 1. default `ofi+tcp` -> `ofi+tcp;ofi_rxm` in launch.py
> 2. update the provider testing branch to use `ofi+tcp;ofi_rxm`
> 3. remove the default `ofi+tcp` -> `ofi+tcp;ofi_rxm` in launch.py

I'm not sure we need to go that much trouble as long as we're able to update the corresponding `provider-*-testing` branch before it runs over the weekend after the master/release branch changes land.  That said this `provider-*-testing` branch could/should be run as a parallel PR - and could serve as the verification of the tcp provider change, possibly negating the need for this PR to be run with multiple providers.

Comment (kjacque) [reply src/tests/ftest/launch.py:64]: > We should test this PR with different `Test-provider: <proivider>` commit pragmas (or build params) and include at least `test_dmg_network_scan_basic` along with the `pr` tag.

Will do. I need to re-run anyway to skip some tests that are failing due to a different bug.

**Comment by kjacque** [comment]
Reverting to a draft while we run the various testing that needs to happen.

**Comment by kjacque** [comment]
The test failure is unrelated. https://daosio.atlassian.net/browse/DAOS-13412

**Comment by kjacque** [comment]
We need to keep in mind the provider branches once this PR lands. Whenever we are approved to land, we should wait until a Monday to actually land, so as to provide time to update the provider branches before a weekend run.

**Review by knard-intel — DISMISSED** [review]

**Review by tanabarr — APPROVED** [review]

**Comment by kjacque** [comment]
The daos_extend_simple test failure on verbs is https://daosio.atlassian.net/browse/DAOS-13412

**Review by daltonbohning — APPROVED** [review]

**Review by soumagne — APPROVED** [review]

**Comment by mchaarawi** [comment]
i feel like this should be documented somewhere now because the with / without rxm matrix has grown and diverged between verbs and tcp. this can be done in a separate PR, although would prefer in the same PR.

i will leave this for @phender to land since there is a comment about provider branch testing.

**Review by daltonbohning — CHANGES_REQUESTED** [review]
@kjacque After talking with @phender, it seems safer to verify provider-testing-tcp before merging this. After https://github.com/daos-stack/daos/pull/12601 is successful, we can land this PR and that one bakc-to-back. It only needs to run VM, so it should be quick.
-1 just until the provider PR passes

**Review by phender — CHANGES_REQUESTED** [review]
Because pipeline-lib uses `'ofi+tcp'` as the default launch.py argument in CI for *all* branches, we'll need a workaround to be able to define a different value per branch until https://daosio.atlassian.net/browse/DAOS-13934 can be implemented.  I'd suggest the following change to launch.py:

```
diff --git a/src/tests/ftest/launch.py b/src/tests/ftest/launch.py
index e8503f83a3..4e3ccf723c 100755
--- a/src/tests/ftest/launch.py
+++ b/src/tests/ftest/launch.py
@@ -66,6 +66,10 @@ PROVIDER_KEYS = OrderedDict(
         ("opx", "ofi+opx"),
     ]
 )
+# Temporary pipeline-lib workaround until DAOS-13934 is implemented
+PROVIDER_ALIAS = {
+    "ofi+tcp": "ofi+tcp;ofi_rxm"
+}
 PROCS_TO_CLEANUP = [
     "daos_server", "daos_engine", "daos_agent", "cart_ctl", "orterun", "mpirun", "dfuse"]
 TYPES_TO_UNMOUNT = ["fuse.daos"]
@@ -1282,6 +1286,8 @@ class Launch():
         """
         logger.debug("-" * 80)
         # Use the detected provider if one is not set
+        if provider in PROVIDER_ALIAS:
+            provider = PROVIDER_ALIAS[provider]
         if not provider:
             provider = os.environ.get("CRT_PHY_ADDR_STR")
         if provider is None:

```

File: src/tests/ftest/launch.py
Comment (phender) [inline src/tests/ftest/launch.py:1290]: To appease pylint:
```suggestion
        provider = PROVIDER_ALIAS.get(provider, provider)
```

**Review by phender — APPROVED** [review]

**Review by soumagne — APPROVED** [review]

**Review by daltonbohning — APPROVED** [review]

**Pull request merged**
